src/pybind/mgr/cephadm/module.py

```diff
@@ -2875,68 +2875,145 @@ class HostAssignment(object):
     """
 
     def __init__(self,
-                 spec=None,  # type: Optional[orchestrator.ServiceSpec]
+                 spec,  # type: orchestrator.ServiceSpec
+                 get_hosts_func,  # type: Callable[[Optional[str]],List[str]]
+                 get_daemons_func, # type: Callable[[str],List[orchestrator.DaemonDescription]]
+
                  scheduler=None,  # type: Optional[BaseScheduler]
-                 get_hosts_func=None,  # type: Optional[Callable[[Optional[str]],List[str]]]
-                 service_type=None,  # type: Optional[str]
                  ):
-        assert spec and get_hosts_func and service_type
+        assert spec and get_hosts_func and get_daemons_func
         self.spec = spec  # type: orchestrator.ServiceSpec
         self.scheduler = scheduler if scheduler else SimpleScheduler(self.spec.placement)
         self.get_hosts_func = get_hosts_func
-        self.daemon_type = service_type
+        self.get_daemons_func = get_daemons_func
+        self.service_name = spec.service_name()
 
     def load(self):
         # type: () -> orchestrator.ServiceSpec
         """
         Load hosts into the spec.placement.hosts container.
         """
-        self.load_labeled_hosts()
-        self.assign_hosts()
+        # respect any explicit host list
+        if self.spec.placement.hosts and not self.spec.placement.count:
+            logger.info('Provided hosts: %s' % self.spec.placement.hosts)
+            if not self.spec.placement.count:
+                self.spec.placement.count = len(self.spec.placement.hosts)
+            return self.spec
+
+        # respect all_hosts=true
+        if self.spec.placement.all_hosts:
+            candidates = [
+                HostPlacementSpec(x, '', '')
+                for x in self.get_hosts_func(None)
+            ]
+            logger.info('All hosts: {}'.format(candidates))
+            self.spec.placement.set_hosts(candidates)
+            self.spec.placement.count = len(self.spec.placement.hosts)
+            return self.spec
+
+        if self.spec.placement.hosts and \
+           self.spec.placement.count and \
+           len(self.spec.placement.hosts) >= self.spec.placement.count:
+            # use the provided hosts and our candidates
+            hosts = self.spec.placement.hosts
+        else:
+            # select candidate hosts based on labels, types, etc.
+            hosts = self.pick_candidates()
+
+        if not self.spec.placement.count:
+            logger.info('Labeled hosts: {}'.format(hosts))
```

Comment:
Those are not necessarily 'labeled' hosts. The conditional's if branch from above assigns explicitly set hosts to `hosts` if `self.placement.hosts` is populated.

Reply:
the first branch is only taken if count is true.  it's convoluted, i know.. i was struggling to make is clearer, but it's hard when all of the fields have slightly different meanings when used in combination with each other.  i think i covered all of the cases in the tests

Reply:
yep, the tests certainly helped!